File: Scripts/Perl/tile.pl
# Tile current selection with pattern inside selection.
# Author: Andrew Trevorrow (andrew@trevorrow.com), June 2007.
# Updated to handle multi-state patterns, Aug 2008.

use strict;

my @selrect = g_getselrect();
g_exit("There is no selection.") if @selrect == 0;

my $selpatt = g_getcells(@selrect);
g_exit("No pattern in selection.") if @{$selpatt} == 0;

# determine if selpatt is one-state or multi-state
my $inc = 2;
if (@{$selpatt} & 1 == 1) { $inc = 3 }

# ------------------------------------------------------------------------------

# return a rect which is the minimal bounding box of given pattern
sub getminbox {
    my $cells = shift;
    my $len = @{$cells};
    return () if $len < 2;
    
    my $minx = $cells->[0];
    my $miny = $cells->[1];
    my $maxx = $minx;
    my $maxy = $miny;
    
    # ignore padding int if present
    $len -= 1 if ($inc == 3) and ($len % 3 == 1);
    
    for (my $x = 0; $x < $len; $x += $inc) {
        if ($cells->[$x] < $minx) { $minx = $cells->[$x] }
        if ($cells->[$x] > $maxx) { $maxx = $cells->[$x] }
    }
    for (my $y = 1; $y < $len; $y += $inc) {
        if ($cells->[$y] < $miny) { $miny = $cells->[$y] }
        if ($cells->[$y] > $maxy) { $maxy = $cells->[$y] }
    }
    
    return ($minx, $miny, $maxx - $minx + 1, $maxy - $miny + 1);
}

# ------------------------------------------------------------------------------

sub clip_left {
    my ($cells, $left) = @_;
    my $len = @{$cells};
    my $x = 0;
    if ($inc == 3) {
        #  ignore padding int if present
        $len -= 1 if $len % 3 == 1;
        while ($x < $len) {
            if ($cells->[$x] >= $left) {
                g_setcell($cells->[$x], $cells->[$x+1], $cells->[$x+2]);
            }
            $x += 3;
        }
    } else {
        while ($x < $len) {
            if ($cells->[$x] >= $left) {
                g_setcell($cells->[$x], $cells->[$x+1], 1);
            }
            $x += 2;
        }
    }
}

# ------------------------------------------------------------------------------

sub clip_right {
    my ($cells, $right) = @_;
    my $len = @{$cells};
    my $x = 0;
    if ($inc == 3) {
        #  ignore padding int if present
        $len -= 1 if $len % 3 == 1;
        while ($x < $len) {
            if ($cells->[$x] <= $right) {
                g_setcell($cells->[$x], $cells->[$x+1], $cells->[$x+2]);
            }
            $x += 3;
        }
    } else {   
        while ($x < $len) {
            if ($cells->[$x] <= $right) {
                g_setcell($cells->[$x], $cells->[$x+1], 1);
            }
            $x += 2;
        }
    }
}

# ------------------------------------------------------------------------------

sub clip_top {
    my ($cells, $top) = @_;
    my $len = @{$cells};
    my $y = 1;
    if ($inc == 3) {
        #  ignore padding int if present
        $len -= 1 if $len % 3 == 1;
        while ($y < $len) {
            if ($cells->[$y] >= $top) {
                g_setcell($cells->[$y-1], $cells->[$y], $cells->[$y+1]);
            }
            $y += 3;
        }
    } else {   
        while ($y < $len) {
            if ($cells->[$y] >= $top) {
                g_setcell($cells->[$y-1], $cells->[$y], 1);
            }
            $y += 2;
        }
    }
}

# ------------------------------------------------------------------------------

sub clip_bottom {
    my ($cells, $bottom) = @_;
    my $len = @{$cells};
    my $y = 1;
    if ($inc == 3) {
        #  ignore padding int if present
        $len -= 1 if $len % 3 == 1;
        while ($y < $len) {
            if ($cells->[$y] <= $bottom) {
                g_setcell($cells->[$y-1], $cells->[$y], $cells->[$y+1]);
            }
            $y += 3;
        }
    } else {   
        while ($y < $len) {
            if ($cells->[$y] <= $bottom) {
                g_setcell($cells->[$y-1], $cells->[$y], 1);
            }
            $y += 2;
        }
    }
}

# ------------------------------------------------------------------------------

# set selection edges
my $selleft = $selrect[0];
my $seltop = $selrect[1];
my $selright = $selleft + $selrect[2] - 1;
my $selbottom = $seltop + $selrect[3] - 1;

# find selpatt's minimal bounding box
my @bbox = getminbox($selpatt);
my $i;

# first tile selpatt horizontally, clipping where necessary
my $left = $bbox[0];
my $right = $left + $bbox[2] - 1;
$i = 0;
while ($left > $selleft) {
    $left -= $bbox[2];
    $i += 1;
    if ($left >= $selleft) {
        g_putcells($selpatt, -$bbox[2] * $i, 0);
    } else {
        my $tempcells = g_transform($selpatt, -$bbox[2] * $i, 0);
        clip_left($tempcells, $selleft);
    }
}
$i = 0;
while ($right < $selright) {
    $right += $bbox[2];
    $i += 1;
    if ($right <= $selright) {
        g_putcells($selpatt, $bbox[2] * $i, 0);
    } else {
        my $tempcells = g_transform($selpatt, $bbox[2] * $i, 0);
        clip_right($tempcells, $selright);
    }
}

# get new selection pattern and tile vertically, clipping where necessary
$selpatt = g_getcells(@selrect);
@bbox = getminbox($selpatt);
my $top = $bbox[1];
my $bottom = $top + $bbox[3] - 1;
$i = 0;
while ($top > $seltop) {
    $top -= $bbox[3];
    $i += 1;
    if ($top >= $seltop) {
        g_putcells($selpatt, 0, -$bbox[3] * $i);
    } else {
        my $tempcells = g_transform($selpatt, 0, -$bbox[3] * $i);
        clip_top($tempcells, $seltop);
    }
}
$i = 0;
while ($bottom < $selbottom) {
    $bottom += $bbox[3];
    $i += 1;
    if ($bottom <= $selbottom) {
        g_putcells($selpatt, 0, $bbox[3] * $i);
    } else {
        my $tempcells = g_transform($selpatt, 0, $bbox[3] * $i);
        clip_bottom($tempcells, $selbottom);
    }
}

g_fitsel() if !g_visrect(@selrect);

File: Scripts/Lua/tile.lua
-- Tile current selection with pattern inside selection.
-- Author: Andrew Trevorrow (andrew@trevorrow.com), Mar 2016.

local g = golly()
local gp = require "gplus"

local selrect = g.getselrect()
if #selrect == 0 then g.exit("There is no selection.") end

local selpatt = g.getcells(selrect)
if #selpatt == 0 then g.exit("No pattern in selection.") end

-- determine if selpatt is one-state or multi-state
local inc = 2
if (#selpatt & 1) == 1 then inc = 3 end

--------------------------------------------------------------------------------

local function clip_left(cells, left)
    local len = #cells
    local x = 1
    if inc == 3 then
        -- ignore padding int if present
        if len % 3 == 1 then len = len - 1 end
        while x <= len do
            if cells[x] >= left then
                g.setcell(cells[x], cells[x+1], cells[x+2])
            end
            x = x + 3
        end
    else
        while x <= len do
            if cells[x] >= left then
                g.setcell(cells[x], cells[x+1], 1)
            end
            x = x + 2
        end
    end
end

--------------------------------------------------------------------------------

local function clip_right(cells, right)
    local len = #cells
    local x = 1
    if inc == 3 then
        -- ignore padding int if present
        if len % 3 == 1 then len = len - 1 end
        while x <= len do
            if cells[x] <= right then
                g.setcell(cells[x], cells[x+1], cells[x+2])
            end
            x = x + 3
        end
    else   
        while x <= len do
            if cells[x] <= right then
                g.setcell(cells[x], cells[x+1], 1)
            end
            x = x + 2
        end
    end
end

--------------------------------------------------------------------------------

local function clip_top(cells, top)
    local len = #cells
    local y = 2
    if inc == 3 then
        -- ignore padding int if present
        if len % 3 == 1 then len = len - 1 end
        while y <= len do
            if cells[y] >= top then
                g.setcell(cells[y-1], cells[y], cells[y+1])
            end
            y = y + 3
        end
    else   
        while y <= len do
            if cells[y] >= top then
                g.setcell(cells[y-1], cells[y], 1)
            end
            y = y + 2
        end
    end
end

--------------------------------------------------------------------------------

local function clip_bottom(cells, bottom)
    local len = #cells
    local y = 2
    if inc == 3 then
        -- ignore padding int if present
        if len % 3 == 1 then len = len - 1 end
        while y <= len do
            if cells[y] <= bottom then
                g.setcell(cells[y-1], cells[y], cells[y+1])
            end
            y = y + 3
        end
    else   
        while y <= len do
            if cells[y] <= bottom then
                g.setcell(cells[y-1], cells[y], 1)
            end
            y = y + 2
        end
    end
end

--------------------------------------------------------------------------------

-- get selection edges
local selleft, seltop, selright, selbottom = gp.getedges(selrect)

-- find selpatt's minimal bounding box
local bbox = gp.getminbox(selpatt)
local i

-- first tile selpatt horizontally, clipping where necessary
local left = bbox.x
local right = left + bbox.wd - 1
i = 0
while left > selleft do
    left = left - bbox.wd
    i = i + 1
    if left >= selleft then
        g.putcells(selpatt, -bbox.wd * i, 0)
    else
        local tempcells = g.transform(selpatt, -bbox.wd * i, 0)
        clip_left(tempcells, selleft)
    end
end
i = 0
while right < selright do
    right = right + bbox.wd
    i = i + 1
    if right <= selright then
        g.putcells(selpatt, bbox.wd * i, 0)
    else
        local tempcells = g.transform(selpatt, bbox.wd * i, 0)
        clip_right(tempcells, selright)
    end
end

-- get new selection pattern and tile vertically, clipping where necessary
selpatt = g.getcells(selrect)
bbox = gp.getminbox(selpatt)
local top = bbox.y
local bottom = top + bbox.ht - 1
i = 0
while top > seltop do
    top = top - bbox.ht
    i = i + 1
    if top >= seltop then
        g.putcells(selpatt, 0, -bbox.ht * i)
    else
        local tempcells = g.transform(selpatt, 0, -bbox.ht * i)
        clip_top(tempcells, seltop)
    end
end
i = 0
while bottom < selbottom do
    bottom = bottom + bbox.ht
    i = i + 1
    if bottom <= selbottom then
        g.putcells(selpatt, 0, bbox.ht * i)
    else
        local tempcells = g.transform(selpatt, 0, bbox.ht * i)
        clip_bottom(tempcells, selbottom)
    end
end

if not g.visrect(selrect) then g.fitsel() end
